src/lib/Tisch/Internal/Create.hs

```diff
@@ -0,0 +1,113 @@
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE KindSignatures        #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedLabels      #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE PolyKinds             #-}
+{-# LANGUAGE QuasiQuotes           #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+
+-- | This module deals with generating database creation statements from a
+-- Tisch database description.
+module Tisch.Internal.Create
+ ( makeCreateStatement
+ ) where
+
+import           Data.Proxy
+import           Data.Singletons
+import           Data.Text                     as T
+import           GHC.TypeLits
+import qualified Opaleye.PGTypes               as O
+import           Text.InterpolatedString.Perl6
+import           Tisch
+import           Tisch.Internal.Table
+
+-- | Generate a CREATE statement for a given table
+makeCreateStatement
+  :: forall table
+   . ( Tisch.Internal.Table.All ColumnInfo (Columns table)
+     , ConsumeTypeList (Columns table)
+     , KnownSymbol (TableName table)
+     )
+  => Table table
+  -> Text
+makeCreateStatement = colInfosCreateStatement (pack tableName) . colInfos
+  where
+    tableName = symbolVal (Proxy @(TableName table))
+
+-- | Get the term-level information about a table
+colInfos
+  :: forall table
+   . ( Tisch.Internal.Table.All ColumnInfo (Columns table)
+     , ConsumeTypeList (Columns table)
+     )
+  => Table table
+  -> [ColInfo]
+colInfos _ = dis (Proxy @ColumnInfo) colInfo (Proxy @(Columns table))
+
+-- TODO: Support default and read only
+-- | Get a create statement from a description of a table.
+colInfosCreateStatement :: Text -> [ColInfo] -> Text
+colInfosCreateStatement tableName is
+  = [qc|CREATE TABLE "{tableName}" ({intercalate "," args});|]
+  where
+    args = colInfoCreateArg <$> is
+    colInfoCreateArg :: ColInfo -> Text
+    colInfoCreateArg (Column name _wcap rcap unique pgtype ()) =
+      let uniqueSuffix :: Text
+          uniqueSuffix = case unique of
+                           NotUnique -> ""
+                           Unique    -> " UNIQUE"
+          nnSuffix :: Text
+          nnSuffix = case rcap of
+                       R  -> " NOT NULL"
+                       RN -> ""
+      in [qc|"{name}" {unPGType pgtype}{uniqueSuffix}{nnSuffix}|]
+
+-- | A handy synonym for a term level column info represenation
+type ColInfo = Column Text WCap RCap Unique PGType ()
+
+-- | A class to reify a Column
+class ColumnInfo col where
+  colInfo :: proxy col -> ColInfo
+
+instance (KnownSymbol name, SingI wcap, SingI rcap, SingI unique, O.IsSqlType (PgType pgtype))
+  => ColumnInfo ('Column name (wcap :: WCap) (rcap :: RCap) (unique :: Unique) pgtype hsType) where
+  colInfo _ = Column (pack . symbolVal $ Proxy @name)
```

Comment:
You're right that a regular function should do, however this is required in order to pass the constraints to `ConsumeTypeList` as a type with kind `k -> Constraint`.